tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+		if (strcmp(compare, name) == 0) return symbol;
+		symbol = symbol->next;
+	}
+
+	return NULL;
+}
+
+int parse_address_bank(char *buffer_input){
+	int bank;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_bank;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4) {
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the bank part
+	bank = strtol(buffer, &endptr_bank, 16);
+	if (endptr_bank != buffer + strlen(buffer)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return bank;
+}
+
+int parse_address_address(char *buffer_input){
+	int address;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_address;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4){
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the address part
+	address = strtol(buffer_address, &endptr_address, 16);
+	if (endptr_address != buffer_address + strlen(buffer_address)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return address;
+}
+
+int parse_abs_offset(int bank, int address, char type)
+{
+	int offset;
+
+	// Calculate the absolute offset in the ROM
+	if (type == 'w') {
+		if (bank == 0) {
+			offset = address;
+		} else {
+			offset = bank * 0xd000 + (address - 0xd000);
+		}
+	} else if (type == 's') {
+		if (bank == 0) {
+			offset = address;
+		} else {
+			offset = bank * 0xc000 + (address - 0xc000);
+		}
+	} else if (type == 'h') {
+		offset = address;
+	} else {
+		if (bank == 0) {
+			offset = address;
+		} else {
+			offset = bank * 0x4000 + (address - 0x4000);
+		}
+	}
+
+	return offset;
+}
+
+int parse_offset(int offset, char type)
+{
+	int address;
+
+	if (type == 'w') {
+		if ((offset - 0xd000) <= 0) {
+			return offset;
+		} else {
+			address = offset % 0xd000;
+			address = address + 0xd000;
+			return address;
+		}
+	} else if (type == 's') {
+		if ((offset - 0xc000) <= 0) {
+			return offset;
+		} else {
+			address = offset % 0xc000;
+			address = address + 0xc000;
+			return address;
+		}
+	} else if (type == 'h') {
+		return offset;
+	} else {
+		if ((offset - 0x4000) <= 0) {
+			return offset;
+		} else {
+			address = offset % 0x4000;
+			address = address + 0x4000;
+			return address;
+		}
+	}
+	return -1;
+}
+
+struct symbol *parse_symfile(FILE *file)
+{
+	struct symbol *symbols = NULL;
```

Comment:
This is C17, not C89; variables don't all have to be declared at the start of a function, but should be declared when they're needed, in as narrow a scope as possible.